components/tidb_query_codegen/src/rpn_function.rs

```diff
@@ -475,12 +492,65 @@ impl ValidatorFnGenerator {
     }
 }
 
+fn generate_init_metadata_fn(
+    metadata_ctor: &Option<TokenStream>,
+    impl_generics: &ImplGenerics<'_>,
+    where_clause: Option<&WhereClause>,
+) -> TokenStream {
+    let fn_body = if let Some(metadata_ctor) = metadata_ctor {
+        quote! { Box::new(#metadata_ctor(expr)) }
+    } else {
+        quote! { Box::new(()) }
+    };
+    quote! {
+        fn init_metadata #impl_generics (expr: &mut ::tipb::Expr)
+            -> Box<dyn std::any::Any + Send> #where_clause {
+            #fn_body
+        }
+    }
+}
+
+fn generate_downcast_metadata(has_metadata: bool) -> TokenStream {
+    if has_metadata {
+        quote! {
+            let metadata = std::any::Any::downcast_ref(metadata).expect("downcast metadata error");
+        }
+    } else {
+        quote! {}
+    }
+}
+
+fn generate_metadata_type_checker(
+    metadata_ctor: &Option<TokenStream>,
+    impl_generics: &ImplGenerics<'_>,
+    where_clause: Option<&WhereClause>,
+    fn_body: impl FnOnce(&TokenStream) -> TokenStream,
+) -> TokenStream {
+    if let Some(metadata_ctor) = metadata_ctor {
+        let body = fn_body(metadata_ctor);
+        quote! {
+            fn _type_checker #impl_generics (
```

Comment:
will this cause name conflict?

Reply:
Ah, yes. I didn't notice it would for `NormalRpnFn`. Addressed using @andylokandy 's trick below.